### PR TITLE
fix: review the pub types/fields(post-split)(1)

### LINKED ISSUE #259: [post frontend-backend split] Review pub types / fields
With the frontend-backend split, I made a lot of types/fields `pub` to simplify the refactoring process.  We need to review which ones don't need to be `pub`; and or add getter/setters where appropriate to keep a similar style to the original halo2 codebase.

Blocked by https://github.com/privacy-scaling-explorations/halo2/pull/254

### PR DESCRIPTION
**Pull request by duguorong009**
## Description
- review & correct the pub types/fields in every crate

## Related issues
- Resolve #259 

## Changes
- add `(crate)` limit to needed types/fields in every crate
- remove the unused methods
- add `#[allow(dead_code)]` for only necessary case, like `Debug`-derived structs(PinnedVerificationKey, ...)

### CODE REVIEW

File: halo2_backend/src/helpers.rs [hunk 1]
```diff
@@ -105,7 +105,8 @@ impl<F: PrimeField + SerdeObject> SerdePrimeField for F {}
 /// Convert a slice of `bool` into a `u8`.
 ///
 /// Panics if the slice has length greater than 8.
-pub fn pack(bits: &[bool]) -> u8 {
+#[allow(dead_code)]
+pub(crate) fn pack(bits: &[bool]) -> u8 {
```
Comment (ed255) [inline halo2_backend/src/helpers.rs:109]: If this is not `pub` but needs `#[allow(dead_code)]` I think it means that this function is not used internally, but also not exposed publicly.  So maybe it's better to remove it?

I think this was only used here https://github.com/privacy-scaling-explorations/halo2/blob/73408a140737d8336490452193b21f5a7a94e7de/halo2_proofs/src/plonk.rs#L100

The same applies to `unpack`.

File: halo2_backend/src/plonk.rs [hunk 2]
```diff
@@ -230,22 +230,24 @@ impl<C: CurveAffine> VerifyingKey<C> {
...
+    pub(crate) fn fixed_commitments(&self) -> &Vec<C> {
         &self.fixed_commitments
     }
 
     /// Returns `VerifyingKey` of permutation
-    pub fn permutation(&self) -> &permutation::VerifyingKey<C> {
+    #[allow(dead_code)]
+    pub(crate) fn permutation(&self) -> &permutation::VerifyingKey<C> {
         &self.permutation
     }
```
Comment (ed255) [inline halo2_backend/src/plonk.rs:243]: Following my last comment: I would either remove these methods, or keep them `pub`.

File: halo2_backend/src/poly.rs [hunk 3]
```diff
@@ -232,7 +232,8 @@ impl<'a, F: Field, B: Basis> Sub<&'a Polynomial<F, B>> for Polynomial<F, B> {
 
 impl<F: Field> Polynomial<F, LagrangeCoeff> {
     /// Rotates the values in a Lagrange basis polynomial by `Rotation`
-    pub fn rotate(&self, rotation: Rotation) -> Polynomial<F, LagrangeCoeff> {
+    #[allow(dead_code)]
```
Comment (ed255) [inline halo2_backend/src/poly.rs:235]: :suspect:

File: halo2_frontend/src/circuit/floor_planner/single_pass.rs [hunk 4]
```diff
@@ -40,7 +40,7 @@ impl FloorPlanner for SimpleFloorPlanner {
 }
 
 /// A [`Layouter`] for a single-chip circuit.
-pub struct SingleChipLayouter<'a, F: Field, CS: Assignment<F> + 'a> {
+pub(crate) struct SingleChipLayouter<'a, F: Field, CS: Assignment<F> + 'a> {
```
Comment (ed255) [inline halo2_frontend/src/circuit/floor_planner/single_pass.rs:43]: I think this needs to be `pub` for users who want to use this `Layouter`

**Review by ed255 — COMMENTED** [review]
In general I think there should be no `#[allow(dead_code)]` anywhere, because it's only needed for non-`pub` entries, but if they're not public and they're not used anywhere, why keep them?

File: halo2_backend/src/helpers.rs (hunk 1 above)
Comment (ed255) [reply halo2_backend/src/helpers.rs:109]: Exactly.  Actually I'm not sure why this was there before.  Previously, for fixed columns only the commitment was stored; and for selectors it was the assignment that was stored.

Comment (duguorong009) [reply halo2_backend/src/helpers.rs]: Done https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/dcb667a4e4b1e168fa9fb10b77405bf1edd75d04

File: halo2_backend/src/poly.rs (hunk 3 above)
Comment (duguorong009) [reply halo2_backend/src/poly.rs]: Done https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/dcb667a4e4b1e168fa9fb10b77405bf1edd75d04

File: halo2_frontend/src/circuit/floor_planner/single_pass.rs (hunk 4 above)
Comment (duguorong009) [reply halo2_frontend/src/circuit/floor_planner/single_pass.rs]: Done https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/dcb667a4e4b1e168fa9fb10b77405bf1edd75d04

File: halo2_middleware/src/expression.rs [hunk 5]
```diff
@@ -124,7 +126,7 @@ impl<F: Field, V: Variable> Expression<F, V> {
     }
 
     /// Square this expression.
-    pub fn square(self) -> Self {
+    pub(crate) fn square(self) -> Self {
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply halo2_middleware/src/expression.rs]: Done https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/dcb667a4e4b1e168fa9fb10b77405bf1edd75d04

File: halo2_middleware/src/expression.rs [hunk 6]
```diff
@@ -94,7 +95,8 @@ impl<F: Field, V: Variable> Expression<F, V> {
     /// Identifier for this expression. Expressions with identical identifiers
     /// do the same calculation (but the expressions don't need to be exactly equal
     /// in how they are composed e.g. `1 + 2` and `2 + 1` can have the same identifier).
-    pub fn identifier(&self) -> String {
+    #[allow(dead_code)]
+    pub(crate) fn identifier(&self) -> String {
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply halo2_middleware/src/expression.rs]: Done https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/dcb667a4e4b1e168fa9fb10b77405bf1edd75d04

File: halo2_backend/src/poly.rs [hunk 7]
```diff
@@ -27,8 +27,8 @@ pub mod kzg;
 #[cfg(test)]
 mod multiopen_test;
 
-pub use domain::*;
-pub use query::{ProverQuery, VerifierQuery};
+pub(crate) use domain::*;
+pub(crate) use query::{ProverQuery, VerifierQuery};
 pub use strategy::{Guard, VerificationStrategy};
 
 // TODO: move everything from the poly module to the backend.  This requires that the frontend
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply halo2_backend/src/poly.rs]: Done https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/dcb667a4e4b1e168fa9fb10b77405bf1edd75d04

File: halo2_backend/src/plonk.rs (hunk 2 above)
Comment (duguorong009) [reply halo2_backend/src/plonk.rs]: Hmm, how about removing the `permutation` method?
I think `permutation::VerifyingKey` should not be `pub`.

File: halo2_middleware/src/expression.rs [hunk 8]
```diff
@@ -113,7 +115,7 @@ impl<F: Field, V: Variable> Expression<F, V> {
     }
 
     /// Approximate the computational complexity of this expression.
-    pub fn complexity(&self) -> usize {
+    pub(crate) fn complexity(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply halo2_middleware/src/expression.rs]: No, I think it is not used in circuit cost estimation tool.

File: halo2_frontend/src/dev.rs [hunk 9]
```diff
@@ -85,26 +85,31 @@ impl Region {
     }
 
     /// Returns the name of the region.
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply halo2_frontend/src/dev.rs]: Remove the mehods in https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/82ff602840780aebcb845d1266e9a93c5b423785

File: halo2_backend/src/plonk.rs (hunk 2 above)
Comment (duguorong009) [reply halo2_backend/src/plonk.rs]: Remove the `permutation` method in https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/82ff602840780aebcb845d1266e9a93c5b423785

File: halo2_backend/src/plonk/verifier/batch.rs [hunk 10]
```diff
@@ -55,6 +55,7 @@ impl<'params, C: CurveAffine>
     }
 }
 
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (duguorong009) [reply halo2_backend/src/plonk/verifier/batch.rs]: Roll back the `pub` in https://github.com/privacy-scaling-explorations/halo2/pull/307/commits/82ff602840780aebcb845d1266e9a93c5b423785

**Comment by duguorong009** [comment]
Thanks for the guide! @ed255 @davidnevadoc 
I followed the suggestions - roll back some `pub` & remove the unused methods.
Pls re-check the PR.

**Review by ed255 — APPROVED** [review]
LGTM!

**Pull request merged**